### PR TITLE
Homework 10-1

### PR DESCRIPTION
**Pull request by sophiadavis**
Sorry about the diff. Just the last commit has the iterator stuff.

### CODE REVIEW

File: course5/homework06/main.cc [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+    String iteratorTest = "this is a test";
+    String iteratorTest2 = "this is another test";
+    for (auto& elem: iteratorTest) {
+        std::cout << elem << std::endl;
+    }
+
+    std::cout << std::boolalpha;
+    auto any_a = std::any_of(iteratorTest.begin(), iteratorTest.end(), isA);
+    std::cout << any_a << std::endl;
```
Comment (maciekgajewski) [inline course5/homework06/main.cc]: Please ,automate the tests

File: course5/homework06/simple_string.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    char operator[](const int) const;
+    char& operator[](const int);
+
+    void Print() const;
+    int length() const;
+    const char *c_str() const;
+
+    typedef char * iterator;
+    typedef const char * const_iterator;
+    iterator begin() { return &buffer.get()[0]; }
```
Comment (maciekgajewski) [inline course5/homework06/simple_string.h]: Knowing that for any pointer ```p``` and integer ```i```, the following expression are identical:
```p[i]``` === ```*(p + i)```
and
```&(*p)``` === ```p```
And knowing that ```buffer.get()``` is a pointer

How can you simplify this function?

File: course5/homework06/simple_string.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    iterator end() { return &buffer.get()[l]; }
```
Comment (maciekgajewski) [inline course5/homework06/simple_string.h]: I don't think this would work

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
Please, automate your tests, because I don't believe your implementation of end() actually works.

Also - you need a const version of begin and end (+ tests!)

You should also test if you're able to modify the content of your string via iterators.

File: course5/homework06/simple_string.h (hunk 3 above)
Comment (sophiadavis) [reply course5/homework06/simple_string.h]: It seems to be working...? What am I doing wrong?

File: course5/homework06/simple_string.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    char& operator[](const int);
+
+    void Print() const;
+    int length() const;
+    const char *c_str() const;
+
+    typedef char * iterator;
+    typedef const char * const_iterator;
+    iterator begin() { return buffer.get(); }
+    iterator end() { return buffer.get() + l; }
```
Comment (maciekgajewski) [inline course5/homework06/simple_string.h:37]: OK. it is l (small-caps L), not 1 (one). So confusing...
See, ```mLength``` or ```length_``` or ```this->l``` would be much easier to read.

Of course it's good.

File: course5/homework06/simple_string.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    char operator[](const int) const;
+    char& operator[](const int);
+
+    void Print() const;
+    int length() const;
+    const char *c_str() const;
+
+    typedef char * iterator;
+    typedef const char * const_iterator;
+    iterator begin() { return buffer.get(); }
```
Comment (maciekgajewski) [inline course5/homework06/simple_string.h:36]: You also need const begin and end, otherwise you wouldn't be able to iterate over const object.

**Review by maciekgajewski — COMMENTED** [review]
Of course it's good; I was confused by combination of bad typeface and your minimalistic variable names.

**Pull request closed without merging**